src/charm.py

```diff
@@ -236,16 +240,83 @@ def _command(self) -> str:
             args.append("--storage.tsdb.wal-compression")
 
         # Set time series retention time
-        if config.get("metrics_retention_time") and self._is_valid_timespec(
-            config["metrics_retention_time"]
-        ):
-            args.append(f"--storage.tsdb.retention.time={config['metrics_retention_time']}")
+        if self._is_valid_timespec(retention_time := config.get("metrics_retention_time", "")):
+            args.append(f"--storage.tsdb.retention.time={retention_time}")
+
+        # Set time series retention size
+        try:
+            ratio = self._percent_string_to_ratio(config.get("maximum_retention_size", ""))
+
+        except ValueError as e:
+            logger.warning(e)
+            self.unit.status = BlockedStatus(f"Invalid retention size: {e}")
+
+        else:
+            # `storage.tsdb.retention.size` uses the legacy binary format, so "GB" and not "GiB"
+            # https://github.com/prometheus/prometheus/issues/10768
+            # For simplicity, always communicate to prometheus in GiB
+            try:
+                capacity = self._convert_k8s_capacity_to_legacy_binary_gigabytes(
+                    self._get_pvc_capacity(), ratio
+                )
+            except ValueError as e:
+                self.unit.status = BlockedStatus(f"Error calculating retention size: {e}")
+            except LightkubeApiError as e:
+                self.unit.status = BlockedStatus(
+                    f"Error calculating retention size "
+                    f"(try running `juju trust` on this application): {e}"
+                )
+            else:
+                logger.debug("Retention size limit set to %s (%s%%)", capacity, ratio * 100)
+                args.append(f"--storage.tsdb.retention.size={capacity}")
 
         command = ["/bin/prometheus"] + args
 
         return " ".join(command)
 
-    def _is_valid_timespec(self, timeval) -> bool:
+    def _convert_k8s_capacity_to_legacy_binary_gigabytes(self, capacity: str, multiplier) -> str:
+        # Reduce possible ambiguity in unit name by adding a trailing 'B'.
+        # (bitmath doesn't accept e.g. "Gi" because it also supports bits.)
+        if not capacity.endswith("B"):
+            capacity += "B"
+
+        # For simplicity, always communicate to prometheus in GiB
+        storage_value = multiplier * bitmath.parse_string(capacity).to_GiB().value
+        return f"{storage_value}GB"
+
+    def _get_pvc_capacity(self) -> str:
+        # Get PVC capacity from kubernetes
+        client = Client()
```

Comment:
Maybe it's a matter of taste, but I would inject `Client` dependency instead of asking `_get_pvc_capacity` to instantiate it. It will be also easier to test.

Reply:
(Discussed and agreed that leaking implementation is also not ideal.)